`NOT` CompositeExpression in order to filter-out values

remi-san opened the pull request:
Some `Comparisons` do not provide the negation of the `Expression` (for example `EQ` has `NEQ`, but `CONTAINS` doesn't have `NCONTAINS`), making it difficult to create complex `Expressions` for use-cases like "I want to retrieve all books from the list where the title **contain** the word 'spider', but only if they **don't contain** the word 'spider-man'."

To do that, you'd have to filter the list with something like `Criteria::create()->where(Criteria::expr()->contains('title', 'Spider')`, store the result, and then filter it again with `Criteria::create()->where(Criteria::expr()->contains('title', 'Spider-man')`, and remove the elements retrieved by the second filter from the first filtered list.

With the  `NOT` expression, you would be able to do it with one simple operation:
```
Criteria::create()->where(
    Criteria::expr()->andX(
        Criteria::expr()->contains('title', 'Spider'),
        Criteria::expr()->not(
            Criteria::expr()->contains('title', 'Spider-man')
        )
    )
);
```

Of course, it would also allow a lot more of complexe cases where we would just want to inverse the result of an expression.

malarzm commented:
@remi-san thanks for the PR! Your changes on its own are looking good (except for PHPStan failure which may be unrelated). Few things that we (as Doctrine) should consider though:

1. ORM will need to support this in https://github.com/doctrine/orm/blob/2.13.x/lib/Doctrine/ORM/Query/QueryExpressionVisitor.php
2. ORM does not have a `NOT` operator so will need to map all non-complex operators to their negated counterparts (`>` -> `<=` etc)

FWIW ODM doesn't transform expressions to MongoDB queries.

greg0ire commented:
About PHPStan, see https://github.com/doctrine/collections/pull/349

malarzm commented:
Cool, one thing less ^^ So a rebase atop `2.0.x` will be fix CI, but we can do that later

greg0ire commented on tests/Common/Collections/Expr/CompositeExpressionTest.php:
```diff
@@ -14,21 +14,23 @@
-    /** @dataProvider invalidDataProvider */
-    public function testExceptions(string|Value $expression): void
+    /**
+     * @param array{type:string, expressions: mixed} $expressions
```
```suggestion
     * @param list<mixed> $expressions
```

greg0ire commented on tests/Common/Collections/Expr/CompositeExpressionTest.php:
```diff
@@ -14,21 +14,23 @@
+     * @param array{type:string, expressions: mixed} $expressions
+     *
+     * @dataProvider invalidDataProvider
+     * */
```
```suggestion
     */
```

greg0ire commented on tests/Common/Collections/Expr/CompositeExpressionTest.php:
```diff
@@ -14,21 +14,23 @@
 class CompositeExpressionTest extends TestCase
 {
-    /** @return list<array{expression: mixed}> */
+    /** @return list<array{type:string, expressions: mixed}> */
```
```suggestion
    /** @return list<array{type:string, expressions: list<mixed>}> */
```

remi-san commented:
Thank you for the feedback.

I rebased my branch so that `phpstan` can stop failing, and fixed the messed up phpdoc (thanks @greg0ire).

@malarzm I was expecting (fearing ? 😨) this kind of following task, I'll try submitting a PR for Doctrine ORM. Is the `NOT` operator not supported in the ORM on purpose or is it a good way to start introducing it?

greg0ire commented on src/ExpressionBuilder.php:
```diff
@@ -30,6 +30,12 @@ public function orX(Expression ...$expressions)
         return new CompositeExpression(CompositeExpression::TYPE_OR, $expressions);
     }
 
+    /** @return CompositeExpression */
+    public function not(Expression $expression)
```
```suggestion
    public function not(Expression $expression): CompositeExpression
```

remi-san commented:
@malarzm  I've created a PR on [`doctrine/orm`](https://github.com/doctrine/orm/pull/10234) as a proof the `NOT` operator can be handled quite easily (a lot more than what I expected).

malarzm commented:
Ah cool! I must have overlooked it while skimming through code yesterday :)

greg0ire commented:
@remi-san since this is a new feature, I've retargeted to 2.1.x.

The PR looks good to me, I think the docs need to be updated though, there should be at least something new in `docs/en/expressions.rst`, but please check other pages. Also, please squash your commits together (let me know if you need help on doing that).

remi-san commented:
@greg0ire Commits have been squashed and doc has been updated, tell me if there's anything else you'd like me to do.

greg0ire commented:
Thanks @remi-san !
